File: calibrate.js
const CALIBRATE_BY = 10;
const buggedPlayers = [];
const headers = ["Name", "ARIMAPP", "LSTMPP", "PP", "AP", "DIFF"];

const ExcelJS = require("exceljs");
const gameInformation = require("./game_information")
const neuralNetwork = require("./neuralNetwork")
const datasetMaker = require("./makeDataset")

const processAllPlayers = false;

const players = {}
let pointsDataSet;

function init() {
    for (const team of gameInformation.TEAMS) {
        players[team] = [];
    }


    datasetMaker.getDataset().then(datasets => {
        pointsDataSet = datasets.points;
        getBasicPlayersTeams();
    })
}

function getBasicPlayersTeams() {
    for (const playerId in pointsDataSet) {
        const playerData = pointsDataSet[playerId];

        players[playerData.team].push(playerData);
    }

    getPlayerPredictions();
}

function getPlayerPredictions() {
    for (const team of gameInformation.TEAMS) {
        players[team].forEach((playerData, index) => {
            console.log(playerData.id, playerData.name);
            if (buggedPlayers.includes(playerData.id)) return;

            let sum = 0;
            let numGames = 0;
            let totalGames = 0;
            const gws = [];
            for (const dataset in playerData) {
                if (!dataset.startsWith("GW")) continue;
                totalGames++;
                gws.push(playerData[dataset])
                const roundNum = parseInt(dataset.replace("GW", ""))
                
                let beginningRound = gameInformation.CURRENT_SEASON_BEGINNING_ROUND;
                if (gameInformation.CURRENT_GAME_WEEK == 1) {
                    beginningRound = gameInformation.CURRENT_SEASON_BEGINNING_ROUND - gameInformation.SEASON_LENGTH - 1;
                }
        
                if (roundNum >= beginningRound) {
                    sum += playerData[dataset].points
                    numGames++;
                }
            }

            if (!processAllPlayers && 
                (totalGames - CALIBRATE_BY < gameInformation.MIN_GAMES || sum < gameInformation.MIN_SEASON_PPG * numGames || numGames < (gameInformation.CURRENT_GAME_WEEK == 1 ? gameInformation.SEASON_LENGTH : gameInformation.CURRENT_GAME_WEEK - 1) * gameInformation.MIN_SEASON_GAME_PERC || totalGames < 2)) {
                console.log("Not min requirements");
                return;
            }        

            const predBy = {"games": [], "next": 0};
            const trainingPlayerData = {};
            let i;
            for (i = 0; i < gws.length - CALIBRATE_BY; i++) {
                trainingPlayerData[`GW${i + 1}`] = gws[i];
            }
            for (; i < gws.length; i++) {
                predBy.games.push(gws[i].diff)
            }

            if (sum <= 0 || gws.length <= CALIBRATE_BY) {
                var arima = 0;
                var lstm = 0;
            } else {
                var arima = neuralNetwork.doARIMA(gws.slice(0, -CALIBRATE_BY).map((val) => val.points), predBy)[0];

                var lstm = neuralNetwork.doLSTM(trainingPlayerData, predBy)[0];

                if (arima / lstm > neuralNetwork.MAX_DIFF || lstm / arima > neuralNetwork.MAX_DIFF) {
                    arima = 0;
                    lstm = 0;
                }
            }
            
            players[team][index] = {
                name: playerData.name,
                gws,
                arima: arima,
                lstm: lstm
            }
        })
    }

    createCalibrationFile();
}

function createCalibrationFile() {
    const workbook = new ExcelJS.Workbook();
    workbook.created = new Date();

    for (const team of gameInformation.TEAMS) {
        if (players[team].filter(player => player.arima != 0 && player.lstm != 0 && player.arima >= -neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.lstm >= -neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.arima <= neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.lstm <= neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.gws.slice(-CALIBRATE_BY).map((obj) => obj.points).reduce((acc, val) => acc + val, 0) >= CALIBRATE_BY * gameInformation.MIN_SEASON_PPG).length == 0) {
            console.log(`Could not find any players for ${team}`);
            continue;
        }
        const sheet = workbook.addWorksheet(team);

        sheet.getCell("H2").value = "ARIMA";
        sheet.getCell("I2").value = 0;

        sheet.getCell("H3").value = "LSTM";
        sheet.getCell("I3").value = 0;

        sheet.getCell("H5").value = "OFF";
        sheet.getCell("I5").value = {formula: `=(SUM(Table${team}[PP]-Table${team}[AP]))^2`, result: -1};
        
        sheet.getCell("H7").value = "AVG";
        sheet.getCell("I7").value = {formula: `=AVERAGE(Table${team}[DIFF])/${CALIBRATE_BY}`, result: -1};

        sheet.addTable({
            name: `Table${team}`,
            ref: "A1",
            headerRow: true,
            style: {
                showRowStripes: true
            },
            columns: headers.map(header => {
                return {
                    name: header,
                    key: header,
                    filterButton: true
                }
            }),
            rows: players[team].filter(player => player.arima != 0 && player.lstm != 0 && player.arima >= -neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.lstm >= -neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.arima <= neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.lstm <= neuralNetwork.RETRY_POINT * CALIBRATE_BY && player.gws.slice(-CALIBRATE_BY).map((obj) => obj.points).reduce((acc, val) => acc + val, 0) >= CALIBRATE_BY * gameInformation.MIN_SEASON_PPG).map(player => {
                return [player.name,
                        player.arima, 
                        player.lstm, 
                        {formula: `=[@ARIMAPP]*$I$2+[@LSTMPP]*$I$3`, result: -1}, 
                        player.gws.slice(-CALIBRATE_BY).map((obj) => obj.points).reduce((acc, val) => acc + val, 0),
                        {formula: `=ABS([@PP]-[@AP])`, result: -1}
                    ];
            })
        });

        sheet.addConditionalFormatting({
            ref: 'I7',
            rules: [
                {
                    type: 'colorScale',
                    cfvo: [
                        {
                            type: 'num',
                            value: 0
                        },
                        {
                            type: 'num',
                            value: 1
                        },
                        {
                            type: 'num',
                            value: 2
                        }
                    ],
                    color: [
                        {
                            argb: '00FF00'
                        },
                        {
                            argb: 'FFFF00'
                        },
                        {
                            argb: 'FF0000'
                        }
                    ]
                }
            ]
        })
    }

    workbook.xlsx.writeFile(`./Calibrations/${gameInformation.CURRENT_SEASON}/Week ${gameInformation.CURRENT_GAME_WEEK}.xlsx`);
}


init();

File: calibrate.py
from ai import do_arima, do_lstm, do_forest


def process_player_data(player_data, current_season_beginning_round, current_game_week, season_length, min_games,
                        min_season_ppg, min_season_game_percentage, calibrate_by, bugged_players, process_all_players,
                        max_diff):
    if player_data['id'] in bugged_players:
        return -1, -1, -1, -1, 0

    points_sum = 0
    num_games = 0
    total_games = 0

    gws = []

    for dataset, data in player_data.items():
        if not dataset.startswith('GW'):
            continue

        round_num = int(dataset.replace("GW", ""))
        if round_num >= current_season_beginning_round + current_game_week - 1:
            continue

        total_games += 1
        gws.append(data)

        beginning_round = current_season_beginning_round
        if current_game_week <= calibrate_by:
            beginning_round = current_season_beginning_round - season_length - 1

        if round_num >= beginning_round:
            points_sum += data['points']
            num_games += 1

    if not process_all_players and (
            total_games - calibrate_by < min_games or points_sum < min_season_ppg * num_games or num_games < (
            season_length if current_game_week <= calibrate_by else current_game_week - 1) *
            min_season_game_percentage):
        return -1, -1, -1, -1, 0

    pred_by = []
    training_player_data = {'id': player_data['id'], 'position': player_data['position'],
                            'first_name': player_data['first_name'], 'last_name': player_data['last_name'],
                            'name': player_data['name'], 'team': player_data['team']}

    for gw_num in range(0, len(gws) - calibrate_by + 1):
        training_player_data[f"GW{gw_num + 1}"] = gws[gw_num]

    actual_points = 0
    for gw in gws[len(gws) - calibrate_by + 1:]:
        pred_by.append(gw['diff'])
        actual_points += gw['points']

    if actual_points <= 0 or len(gws[:-calibrate_by]) <= calibrate_by:
        arima = 0
        lstm = 0
        forest = 0
    else:
        try:
            arima = sum(do_arima(list(map(lambda x: x['points'], gws[:-calibrate_by])), pred_by))
            lstm = sum(do_lstm(training_player_data, pred_by))
            forest = sum(do_forest(training_player_data, pred_by))
        except Exception as e:
            print(e)
            print('AN ERROR HAPPENED')
            arima = 0
            lstm = 0
            forest = 0

        if arima != 0 and lstm != 0 and (arima / lstm > max_diff or lstm / arima > max_diff):
            arima = 0
            lstm = 0
            forest = 0

    return arima, lstm, forest, actual_points, actual_points / len(pred_by)
